File: Android/doraemonkit-plugin/src/main/groovy/com/didichuxing/doraemonkit/plugin/extension/SlowMethodExt.java
package com.didichuxing.doraemonkit.plugin.extension;

import org.gradle.api.Action;
import org.gradle.util.ConfigureUtil;

import java.util.ArrayList;
import java.util.List;

import groovy.lang.Closure;

/**
 * ================================================
 * 作    者：jint（金台）
 * 版    本：1.0
 * 创建日期：2020/4/28-14:56
 * 描    述：
 * 修订历史：
 * ================================================
 */
public class SlowMethodExt {
    public static final int STRATEGY_STACK = 0;
    public static final int STRATEGY_NORMAL = 1;
    /**
     * 0:打印函数调用栈  1:普通模式 运行时打印某个函数的耗时 全局业务代码函数插入
     */
    public int strategy = STRATEGY_STACK;


    /**
     * 函数功能开关
     */
    public boolean methodSwitch = true;

    public void strategy(int strategy) {
        this.strategy = strategy;
    }

    public void methodSwitch(boolean methodSwitch) {
        this.methodSwitch = methodSwitch;
    }


    public StackMethodExt stackMethod = new StackMethodExt();
    public NormalMethodExt normalMethod = new NormalMethodExt();

    public void stackMethod(Closure<StackMethodExt> closure) {

        ConfigureUtil.configure(closure, this.stackMethod);
    }

    public void normalMethod(Closure<NormalMethodExt> closure) {
        ConfigureUtil.configure(closure, this.normalMethod);
    }

    public static class StackMethodExt {
        /**
         * 默认值为5ms
         */
        public int thresholdTime = 5;
        public List<String> enterMethods = new ArrayList<>();

        public void thresholdTime(int thresholdTime) {
            this.thresholdTime = thresholdTime;
        }

        public void normalMethod(List<String> enterMethods) {
            this.enterMethods = enterMethods;
        }

        @Override
        public String toString() {
            return "StackMethodExt{" +
                    "thresholdTime=" + thresholdTime +
                    ", enterMethods=" + enterMethods +
                    '}';
        }
    }


    public static class NormalMethodExt {
        /**
         * 默认值为500ms
         */
        public int thresholdTime = 500;
        public List<String> packageNames = new ArrayList<>();
        public List<String> methodBlacklist = new ArrayList<>();

        public void thresholdTime(int thresholdTime) {
            this.thresholdTime = thresholdTime;
        }


        public void packageNames(List<String> packageNames) {
            this.packageNames = packageNames;
        }

        public void methodBlacklist(List<String> methodBlacklist) {
            this.methodBlacklist = methodBlacklist;
        }

        @Override
        public String toString() {
            return "NormalMethodExt{" +
                    "thresholdTime=" + thresholdTime +
                    ", packageNames=" + packageNames +
                    ", methodBlacklist=" + methodBlacklist +
                    '}';
        }
    }

    @Override
    public String toString() {
        return "SlowMethodExt{" +
                "strategy=" + strategy +
                ", methodSwitch=" + methodSwitch +
                ", stackMethod=" + stackMethod +
                ", normalMethod=" + normalMethod +
                '}';
    }
}

File: Android/buildSrc/src/main/kotlin/com/didichuxing/doraemonkit/plugin/extension/SlowMethodExt.kt
package com.didichuxing.doraemonkit.plugin.extension

import groovy.lang.Closure
import org.gradle.util.ConfigureUtil
import java.util.*

/**
 * ================================================
 * 作    者：jint（金台）
 * 版    本：1.0
 * 创建日期：2020/4/28-14:56
 * 描    述：
 * 修订历史：
 * ================================================
 */

open class SlowMethodExt(
        //0:打印函数调用栈  1:普通模式 运行时打印某个函数的耗时 全局业务代码函数插入
        var strategy: Int = STRATEGY_STACK,
        //函数功能开关
        var methodSwitch: Boolean = true,
        //函数调用栈模式
        var stackMethod: StackMethodExt = StackMethodExt(),
        //普通模式
        var normalMethod: NormalMethodExt = NormalMethodExt()) {


    /**
     * 函数功能开关
     */
    fun strategy(strategy: Int) {
        this.strategy = strategy
    }

    fun methodSwitch(methodSwitch: Boolean) {
        this.methodSwitch = methodSwitch
    }

    fun stackMethod(closure: Closure<StackMethodExt?>?) {
        ConfigureUtil.configure(closure, stackMethod)
    }

    fun normalMethod(closure: Closure<NormalMethodExt?>?) {
        ConfigureUtil.configure(closure, normalMethod)
    }

    class StackMethodExt(
            //默认阈值为5ms
            var thresholdTime: Int = 5,
            //入口函集合
            var enterMethods: MutableList<String> = mutableListOf()) {
        /**
         * 默认值为5ms
         */
        fun thresholdTime(thresholdTime: Int) {
            this.thresholdTime = thresholdTime
        }

        fun normalMethod(enterMethods: MutableList<String>) {
            this.enterMethods = enterMethods
        }

        override fun toString(): String {
            return "StackMethodExt{" +
                    "thresholdTime=" + thresholdTime +
                    ", enterMethods=" + enterMethods +
                    '}'
        }
    }

    class NormalMethodExt(
            //默认阈值为500ms
            var thresholdTime: Int = 500,
            //普通函数的插装包名集合
            var packageNames: MutableList<String> = mutableListOf(),
            //插桩黑名单
            var methodBlacklist: MutableList<String> = mutableListOf()) {
        /**
         * 默认值为500ms
         */

        fun thresholdTime(thresholdTime: Int) {
            this.thresholdTime = thresholdTime
        }

        fun packageNames(packageNames: MutableList<String>) {
            this.packageNames = packageNames
        }

        fun methodBlacklist(methodBlacklist: MutableList<String>) {
            this.methodBlacklist = methodBlacklist
        }

        override fun toString(): String {
            return "NormalMethodExt{" +
                    "thresholdTime=" + thresholdTime +
                    ", packageNames=" + packageNames +
                    ", methodBlacklist=" + methodBlacklist +
                    '}'
        }
    }

    override fun toString(): String {
        return "SlowMethodExt{" +
                "strategy=" + strategy +
                ", methodSwitch=" + methodSwitch +
                ", stackMethod=" + stackMethod +
                ", normalMethod=" + normalMethod +
                '}'
    }

    companion object {
        const val STRATEGY_STACK = 0
        const val STRATEGY_NORMAL = 1
    }
}